Mathlib/Geometry/Manifold/IntegralCurve.lean

```diff
@@ -0,0 +1,153 @@
+/-
+Copyright (c) 2023 Winston Yin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Winston Yin
+-/
+import Mathlib.Analysis.ODE.PicardLindelof
+import Mathlib.Geometry.Manifold.ContMDiff
+import Mathlib.Geometry.Manifold.MFDeriv
+
+/-!
+# Integral curves of vector fields on a manifold
+
+For any continuously differentiable vector field on a manifold `M` and any chosen non-boundary point
+`x₀ : M`, an integral curve `γ : ℝ → M` exists such that `γ t₀ = x₀` and the tangent vector of `γ`
+at `t` coincides with the vector field at `γ t` for all `t` within an open interval around `t₀`.
+
+As a corollary, such an integral curve exists for any starting point `x₀` if `M` is a manifold
+without boundary.
+
+## Tags
+
+integral curve, vector field
+-/
+
+open scoped Manifold
+
+section
+
+variable
+  {𝕜 : Type*} [NontriviallyNormedField 𝕜]
+  {E : Type*} [NormedAddCommGroup E] [NormedSpace 𝕜 E]
+  {H : Type*} [TopologicalSpace H] {I : ModelWithCorners 𝕜 E H}
+  {M : Type*} [TopologicalSpace M] [ChartedSpace H M] [SmoothManifoldWithCorners I M]
+  {E' : Type*} [NormedAddCommGroup E'] [NormedSpace 𝕜 E']
+  {H' : Type*} [TopologicalSpace H'] {I' : ModelWithCorners 𝕜 E' H'}
+  {M' : Type*} [TopologicalSpace M'] [ChartedSpace H' M'] [SmoothManifoldWithCorners I' M']
+
+/-- An interior point of a manifold is a point whose image in the model vector space is in the
+interior of the chart's target. -/
+def ModelWithCorners.IsInteriorPoint (x : M) := extChartAt I x x ∈ interior (extChartAt I x).target
+
+lemma ModelWithCorners.Boundaryless.isOpen_target
+    [I.Boundaryless] {x : M} : IsOpen (extChartAt I x).target := by
+rw [extChartAt_target, ModelWithCorners.Boundaryless.range_eq_univ, Set.inter_univ]
+exact (ModelWithCorners.continuous_symm _).isOpen_preimage _ (LocalHomeomorph.open_target _)
+
+lemma ModelWithCorners.Boundaryless.isInteriorPoint
+    [I.Boundaryless] {x : M} : I.IsInteriorPoint x := by
+rw [ModelWithCorners.IsInteriorPoint,
+  IsOpen.interior_eq ModelWithCorners.Boundaryless.isOpen_target]
+exact LocalEquiv.map_source _ (mem_extChartAt_source _ _)
+
+end
+
+variable
+  {E : Type*} [NormedAddCommGroup E] [NormedSpace ℝ E] [CompleteSpace E]
+  {H : Type*} [TopologicalSpace H] {I : ModelWithCorners ℝ E H}
+  {M : Type*} [TopologicalSpace M] [ChartedSpace H M] [SmoothManifoldWithCorners I M]
+  {E' : Type*} [NormedAddCommGroup E'] [NormedSpace ℝ E']
+  {H' : Type*} [TopologicalSpace H'] {I' : ModelWithCorners ℝ E' H'}
+  {M' : Type*} [TopologicalSpace M'] [ChartedSpace H' M'] [SmoothManifoldWithCorners I' M']
+  {v : (x : M) → TangentSpace I x} {x₀ : M}
+  (hv : ContMDiffAt I I.tangent 1 (fun x => (⟨x, v x⟩ : TangentBundle I M)) x₀) (t₀ : ℝ)
+
+def HasIntegralCurveAt (v : (x : M) → TangentSpace I x) (γ : ℝ → M) (x₀ : M) (t₀ : ℝ) :=
```

Comment:
A first cut at a docstring, can surely be optimised:
```suggestion
/-- If `v : M → TM` is a vector field on `M` and `x : M`, `HasIntegralCurveAt v γ x₀ t₀` means
`γ : ℝ → M` is a differentiable integral curve of `v` with `γ x₀ = t₀. -/
def HasIntegralCurveAt (v : (x : M) → TangentSpace I x) (γ : ℝ → M) (x₀ : M) (t₀ : ℝ) :=

Reply:
Thinking about this, I find `HasIntegralCurveAt` misleading. How about `IsIntegralCurveAt`?

Reply:
I like this. I will also shuffle the arguments around: `IsIntegralCurveAt γ v t₀ x₀`.